### ISSUE TITLE
Datastore: Cannot Implement DynamoDB Version Control Pattern

### ISSUE THREAD

**Issue by sherl0cks** [aws-amplify/amplify-js#7950]
**Describe the bug**
I'm attempting to the [DynamoDB Version Control](https://aws.amazon.com/blogs/database/implementing-version-control-using-amazon-dynamodb/) pattern with Amplify  datastore JS. From what I can tell, this is not possible on the default table index, which requires expensive, unnecessary GSI's to work around datastore's limitations.

My hope is that I can use app sync / data store conflict detection and dynamodb version control history. If datastore is promoting an alternate design pattern to maintain version history (and not just conflict detection), could you please share some documentation?

**To Reproduce**
``` graphql
type VersionControlledType @key(fields: ["id", "version"]) 
{
  id : ID!
  version: String!   #note using Int! seems to break the generated resolvers when using `@connection`
}
```

``` js
const myType = new VersionControlledType({version: 'v1'})
await Datastore.save(myType)
await Datastore.save(VersionControlledType.copy(myType, {updated => { updated.version = 'v2'}))
```

Results in:
``` console
    [WARN] 16:38.425 DataStore {
      localModel: VersionControlledType {
        id: '35117593-fef0-4506-a035-332c78700fe8',
        version: 'v2',
        _version: undefined,
        _lastChangedAt: undefined,
        _deleted: undefined
      },
      message: 'No existing item found in the data source : The conditional request failed (Service: DynamoDb, Status Code: 400, Request ID: A05E6J3LK230HUHTU09CCS3Q3RVV4KQNSO5AEMVJF66Q9ASUAAJG, Extended Request ID: null)',
      operation: 'Update',
      errorType: 'ConditionalCheckFailedException',
      errorInfo: null,
      remoteModel: null
    }
```

**Expected behavior**
'v2' item saved in dynamo. 

**Additional context**

- Datastore is otherwise working as expected, so this does not appear to be an issue with amplify cli configs. 
- Same behavior if I use `@key(fields:[])` to switch the partition key to a different field that `id`. Seems that the sync engine is not aware that the primary key is a concatenation of the partition key and sort key so its trying to update an item instead of writing a new one. Whether that logic is in amplify JS or the generated app sync resolvers, I do not know.

**Comment by sherl0cks** [comment]
After some additional hacking, I am able to get the pattern working but with the following:

``` graphql
type VersionControlledType @key(fields: ["userProvidedId", "version"]) 
{
  userProvidedId : String!
  version: String!   #note using Int! seems to break the generated resolvers 
}
```

Using `amplify codegen models` still generates TS with the `id` field. So I'm guessing datastore has expectations about this value both client side and server side.

``` ts
export declare class VersionControlledType {
  readonly id: string;
  readonly version: string;
  readonly userProvidedId: string;
}
```

Using `VersionControlledType.copy` will carry over the `id` field which is not visible to developers, so we need to use `new` which will give us a new `id`.

``` js

await Datastore.save(new VersionControlledType({userProvidedId: "uniqueId", version: 'v1'}))
await Datastore.save(new VersionControlledType({userProvidedId: "uniqueId", version: 'v2'})))
```

Worth noting that the DynamoDB tables are generated as expected and work as expected.

### Pain

The problem here is that `id` is polluting the dynamodb items / appsync graphql api (which I'd like non datastore clients to query) / client side TS models. Since I cannot use the `id` as an actual unique identifier in my app becauses its really just a datastore implementation details, it's going to be really confusing for other developers.

I'm happy to collaborate on a solution if a maintainer is interested.

**Comment by sherl0cks** [comment]
OK last update here for now. It appears that the Datastore JS cannot generate the correct mutation to update the `v0` record.

``` js 
let latestVersion = new VersionControlledType({userProvidedId: "uniqueId", version: 'v0', data: 'originalValue'})
await Datastore.save(latestVersion)
latestVersion = VersionControlledType.copyOf(latestVersion, updated => {updated.data = 'newValue'})
await Datastore.save(latestVersion)
```

results in 

``` console
    [WARN] 05:22.993 DataStore {
      localModel: SiteModel {
        id: '6ed88758-7ff4-47e1-827f-091272b57831',
        data: 'newValue',
        _version: 1,
        _lastChangedAt: undefined,
        _deleted: undefined
      },
      message: "Variable 'input' has coerced Null value for NonNull type 'String!'",
      operation: 'Update',
      errorType: undefined,
      errorInfo: undefined,
      remoteModel: null
    }
```

Looking at App Sync logs, it appears that Datastore JS is not passing in the required `userProvidedId` and `version` parameters, and this is resulting in the generated vtl resolver to blow up.

My team will not be able to proceed with Datastore for our new product without a fix or workaround to this issue.

**Comment by sherl0cks** [comment]
As a follow up, I am able to implement the pattern using a GSI. This isn't ideal as it will double the storage costs I have for this table, but it does work. My app code is responsible for maintaining the `userProvidedId` across all new items. To do this, the app code always modifies the v0 record, which is used in  `@connections` to other `@models`.  A `versionCount` field 
is kept on the `v0` item, incremented on each `DataStore.save` and then used to  `DataStore.save(new VersionControlledType({userProvidedId: v0.id, version: v0.versionCount` }))`.
```
type VersionControlledType @key(name: "versioned" fields: ["userProvidedId", "version"]) 
{
  id: ID!
  userProvidedId: String!
  version: Int!   # note - connections use id so no issue using Int
  versionCount: Int!
```

**Comment by iartemiev** [comment]
@sherl0cks did you generate your schema via the Amplify CLI? Or have you made changes to it manually, e.g., via the AppSync Console? 

All fields except for `id` should be optional in a model's update mutation input. Only internal fields (`id`, `_version`, `_lastChangedAt`, `_deleted` + any modified user-defined fields should actually be sent. This is required in order to comply with the Automerge conflict resolution strategy.

```js
const latestVersion = await Datastore.save(new VersionControlledType({
  userProvidedId: "uniqueId", 
  version: 'v0', 
  data: 'originalValue'
}));

await Datastore.save(VersionControlledType.copyOf(latestVersion, updated => {updated.data = 'newValue'}))
```

The second save should result in an update mutation, which only includes `data` and the internal fields mentioned above in the input and should not return an error.

Please check `UpdateVersionControlledTypeInput` in the generated schema and see if any fields other than `id` are marked as required.

**Comment by sherl0cks** [comment]
Yes the schema was generated via the amplify cli. I'm intentionally
avoiding going into the appsync schema console.

At this point, I don't have a working reproducer because I moved forward
with the GSI approach, and just accepted that increase in $ for storage is
the tradeoff for the development acceleration of Datastore (which is
tremendous btw - my heartfelt thank you to the entire team). I'll try to
get an OSS reproducer later this week or early next week that I can share.

**Comment by sanyashvets** [comment]
I know this issue
this happens for me when I'm creating new record and then trying to update it almost immediately
the problem is that in this case `_version` is not sending to UpdateRecord mutation and it just.. fails

so, flow in details
- create record (datastore.save(new ...)
- event `outboxMutationEnqueued` fires
- update record (datastore.save(original.copyOf), etc)
* outboxMutationEnqueued* is still processing*
- I see request in network to update record without  `_version` to be sent
- I receive same error. Or `InternalFailure`


IF I WAIT for event `outboxMutationProcessed` to be fired - after it everything updates without any issues

Conflict merge strategy - AutoMerge

**Comment by iartemiev** [comment]
Thank you for the additional info, @sanyashvets, this is the same issue as https://github.com/aws-amplify/amplify-js/issues/7888. We've identified the underlying problem and have opened a [PR](https://github.com/aws-amplify/amplify-js/pull/7935) up that fixes it. Will be merged and deployed soon.

**Comment by sherl0cks** [comment]
Interesting. I had a suspicion that there was a race condition, so I did have a [`sleep`](https://stackoverflow.com/questions/951021/what-is-the-javascript-version-of-sleep) in between my create and update at some point. I'll hold off creating a reproducer until these issues are shipped.

**Comment by sanyashvets** [comment]
yes.. race
between create <-> update / update <-> update
very annoying. IDK what to do. Inserting delays everywhere - I hate myself for it :D :(

**Comment by iartemiev** [comment]
We just deployed changes that should fix this. 
Please install `aws-amplify@3.3.26` or `@aws-amplify/datastore@2.9.15` and let me know if that solves it for you.

**Comment by sanyashvets** [comment]
wait a sec @iartemiev

**Comment by sanyashvets** [comment]
nope, doesn't worked for me

**Comment by iartemiev** [comment]
Can you run this command from the root of your project 
```
npx envinfo --system --binaries --browsers --npmPackages --npmGlobalPackages
```
And then paste the contents here

**Comment by sanyashvets** [comment]
we can chat in AWS Amplify discord channel if it would be better



```
 System:
    OS: macOS 11.2.3
    CPU: (12) x64 Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
    Memory: 94.08 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.16.0 - ~/.nvm/versions/node/v14.16.0/bin/node
    npm: 6.14.11 - ~/.nvm/versions/node/v14.16.0/bin/npm
  Browsers:
    Chrome: 89.0.4389.90
    Firefox: 72.0.2
    Safari: 14.0.3
    Safari Technology Preview: 14.2
  npmPackages:
    @angular-devkit/build-angular: 0.901.7 => 0.901.7 
    @angular/animations: 9.1.9 => 9.1.9 
    @angular/cdk: 9.2.4 => 9.2.4 
    @angular/cli: 9.1.7 => 9.1.7 
    @angular/common: 9.1.9 => 9.1.9 
    @angular/compiler: 9.1.9 => 9.1.9 
    @angular/compiler-cli: 9.1.9 => 9.1.9 
    @angular/core: 9.1.9 => 9.1.9 
    @angular/flex-layout: 9.0.0-beta.31 => 9.0.0-beta.31 
    @angular/forms: 9.1.9 => 9.1.9 
    @angular/language-service: 9.1.9 => 9.1.9 
    @angular/material: 9.2.4 => 9.2.4 
    @angular/material-moment-adapter: 9.2.4 => 9.2.4 
    @angular/platform-browser: 9.1.9 => 9.1.9 
    @angular/platform-browser-dynamic: 9.1.9 => 9.1.9 
    @angular/router: 9.1.9 => 9.1.9 
    @angular/service-worker: 9.1.9 => 9.1.9 
    @angularclass/hmr: 2.1.3 => 2.1.3 
    @auth0/angular-jwt: 4.1.2 => 4.1.2 
    @ionic/angular: 5.1.1 => 5.1.1 
    @ionic/angular-toolkit: ^2.2.0 => 2.3.3 
    @ionic/storage: 2.2.0 => 2.2.0 
    @ngrx/effects: 9.1.2 => 9.1.2 
    @ngrx/entity: 9.1.2 => 9.1.2 
    @ngrx/router-store: 9.1.2 => 9.1.2 
    @ngrx/schematics: 9.1.2 => 9.1.2 
    @ngrx/store: 9.1.2 => 9.1.2 
    @ngrx/store-devtools: 9.1.2 => 9.1.2 
    @ngx-formly/core: 5.5.11 => 5.5.11 
    @ngx-formly/ionic: 5.5.11 => 5.5.11 
    @ngx-formly/material: 5.5.11 => 5.5.11 
    @ngx-formly/schematics: 5.5.11 => 5.5.11 
    @ngx-translate/core: 12.1.2 => 12.1.2 
    @nrwl/angular: 9.3.0 => 9.3.0 
    @nrwl/jest: 9.3.0 => 9.3.0 
    @nrwl/node: 9.3.0 => 9.3.0 
    @nrwl/workspace: 9.3.0 => 9.3.0 
    @tensorflow/tfjs: 1.2.10 => 1.2.10 
    @tensorflow/tfjs-converter: 1.2.10 => 1.2.10 
    @tensorflow/tfjs-core: 1.2.10 => 1.2.10 
    @types/crypto-js: 3.1.47 => 3.1.47 
    @types/faker: 4.1.12 => 4.1.12 
    @types/jasmine: 3.5.10 => 3.5.10 
    @types/jasminewd2: 2.0.8 => 2.0.8 
    @types/jest: 25.1.4 => 25.1.4 
    @types/lodash: 4.14.152 => 4.14.152 
    @types/node: ^12.11.1 => 12.20.6 
    @types/ua-parser-js: 0.7.33 => 0.7.33 
    angular-calendar: 0.28.15 => 0.28.15 
    angular-in-memory-web-api: ^0.11.0 => 0.11.0 
    authorizer-wrapper-el: 1.0.3 => 1.0.3 
    aws-amplify: 3.3.26 => 3.3.26 
    bootstrap: 4.5.0 => 4.5.0 
    classlist.js: 1.1.20150312 => 1.1.20150312 
    codelyzer: ^5.1.2 => 5.2.2 
    comlink: 4.3.0 => 4.3.0 
    core-js: 2.5.7 => 2.5.7 
    crypto-js: 3.1.9-1 => 3.1.9-1 
    faker: 4.1.0 => 4.1.0 
    hammerjs: 2.0.8 => 2.0.8 
    ionic-cache: 5.2.0 => 5.2.0 
    jasmine-core: 3.5.0 => 3.5.0 
    jasmine-marbles: 0.6.0 => 0.6.0 
    jasmine-spec-reporter: 4.2.1 => 4.2.1 
    jest: 25.2.3 => 25.2.3 
    jwt-decode: 2.2.0 => 2.2.0 
    karma: 4.3.0 => 4.3.0 
    karma-chrome-launcher: 3.1.0 => 3.1.0 
    karma-coverage-istanbul-reporter: 2.1.1 => 2.1.1 
    karma-jasmine: 2.0.1 => 2.0.1 
    karma-jasmine-html-reporter: 1.4.2 => 1.4.2 
    konva: 7.1.4 => 7.1.4 
    localforage: ^1.7.3 => 1.9.0 
    lodash: 4.17.19 => 4.17.19 
    mat-currency-format: 0.0.7 => 0.0.7 
    moment: 2.26.0 => 2.26.0 
    ng-mocks: 9.5.0 => 9.5.0 
    ng-pick-datetime: 7.0.0 => 7.0.0 
    ng-zorro-antd: 9.1.2 => 9.1.2 
    ngrx-store-localstorage: ^9.0.0 => 9.0.0 
    ngx-cookie-service: 2.0.0 => 2.0.0 
    ngx-daterangepicker-material: 3.0.1 => 3.0.1 
    ngx-material-timepicker: 5.5.2 => 5.5.2 
    ngx-skeleton-loader: 1.2.2 => 1.2.2 
    ngx-socket-io: 3.0.1 => 3.0.1 
    ngx-swiper-wrapper: 9.0.1 => 9.0.1 
    ngx-toastr: 12.0.1 => 12.0.1 
    object-path: ^0.11.4 => 0.11.5 
    perfect-scrollbar: 1.5.0 => 1.5.0 
    prettier: 2.0.4 => 2.0.4 
    pretty-quick: 2.0.1 => 2.0.1 
    protractor: 5.4.4 => 5.4.4 
    quagga: 0.12.1 => 0.12.1 
    replace-in-file: 6.1.0 => 6.1.0 
    rxjs: 6.5.5 => 6.5.5 
    rxjs-compat: 6.5.5 => 6.5.5 
    subsink: 1.0.1 => 1.0.1 
    swiper: 5.4.1 => 5.4.1 
    ts-jest: 25.2.1 => 25.2.1 
    ts-node: 7.0.1 => 7.0.1 
    tslib: ^1.10.0 => 1.14.1 
    tslint: 5.11.0 => 5.11.0 
    tslint-config-prettier: 1.18.0 => 1.18.0 
    typescript: 3.8.3 => 3.8.3 
    ua-parser-js: ^0.7.21 => 0.7.24 
    uuid: 8.1.0 => 8.1.0 
    web-animations-js: ^2.3.2 => 2.3.2 
    webpack-bundle-analyzer: 3.4.1 => 3.4.1 
    zen-observable-ts: 1.0.0 => 1.0.0 
    zone.js: ~0.10.2 => 0.10.3 
  npmGlobalPackages:
    @aws-amplify/cli: 4.46.0
    npm: 6.14.11

```

**Comment by sanyashvets** [comment]
is there anything I can help with?

**Comment by iartemiev** [comment]
Thank you. If you could provide a repo or sandbox with a minimal reproducible sample demonstrating this issue, that would help a lot

**Comment by sanyashvets** [comment]
right now I cannot, but I can say how to do it. It's easy!

1 - sync with dynamo should be enabled
2 - create record via "DataStore.save(new Record...)"
3 - await record to be created
4 - try update any field and perform it (`DataStore.save(original.copyOf(...)`). You can do it immediately after issue was created OR WITH `SetTimeout() 50-100ms`. At this point `outboxMutationEqueued` event should fire, but it should not be processed yet. As a result, you'll see warning that some required field is missing 

![image](https://user-images.githubusercontent.com/9954697/112538865-fd69cb80-8db8-11eb-8515-c26020e2a650.png)

in network

![image](https://user-images.githubusercontent.com/9954697/112539171-50dc1980-8db9-11eb-8dcf-ecc45c17dee9.png)

param `_version` is missing!
response for that request is:
```
data: null
errorInfo: null
errorType: "ConditionalCheckFailedException"
locations: [{line: 2, column: 3, sourceName: null}]
0: {line: 2, column: 3, sourceName: null}
message: "No existing item found in the data source : The conditional request failed (Service: DynamoDb, Status Code: 400, Request ID: <SOMEID>, Extended Request ID: null)"
```
(of course, there are no `_version` param)

but `await DataStore.save()` returns correct result for but update and create

5 - loop updates in cycle in interval of 50-100ms. You'll see warnings all the time



From my point of view, the fix should be following: `no any update requests (mutations in network tab) shouldn't be fired until outboxMutationProcessed is received and _version param updated in local indexdb`. Because when I do this on my own with `Hub.listen('datastore', () => ....)` - it works good. But with lots of delays because of it, written on my own...  I prefer this solution to be the part of package rather than writing this conditions for every entity...

**Comment by iartemiev** [comment]
Is `original` in your example the result of `DataStore.save`? Or the model instantiation?

The update should be performed like this:
```js
const newPost = await Datastore.save(new Post({
  title: 'New Post'
}));

await Datastore.save(Post.copyOf(newPost, updated => {updated.title = 'Updated'}))
```

**Comment by sanyashvets** [comment]
`original` is result of `DataStore.query(Record, id)`

**Comment by sanyashvets** [comment]
like this
```
const original = await DataStore.query(Post, "123");

await DataStore.save(
  Post.copyOf(original, updated => {
    updated.title = `title ${Date.now()}`;
  })
);
```

**Comment by iartemiev** [comment]
Are you using `Record.copyOf` or `original.copyOf`? A couple of comments back it looked like it may be the 2nd, but it needs to be the first, i.e., `Record.copyOf`

**Comment by sanyashvets** [comment]
like in documentation. `Post.copyOf(original, () => ...)` (`Record.copyOf(original, () => ...)`)

**Comment by iartemiev** [comment]
Ok, I'm AFK at the moment, but I'll put together a sample around the description you provided above later this evening or tomorrow morning. I'll let you know if I was able to reproduce it or if I need more details

**Comment by sanyashvets** [comment]
ok cool. Will wait for it

**Comment by iartemiev** [comment]
I wasn't able to repro that error you posted above with the latest version of amplify. 

I am, however, seeing the unexpected behavior where an update performed right after a create gets overwritten with the values from the create mutation. I'll open a PR to resolve that.

**Comment by sanyashvets** [comment]
yes yes it looks like it
thanks!

**Comment by iartemiev** [comment]
The issue is addressed in aws-amplify@3.3.27.
If you're still experiencing this error in the latest version of Amplify, please let us know and we will re-open this issue and investigate further.
